### PR TITLE
Refactor domain errors

### PR DESCRIPTION
**Pull request by esdrasjnr**
What does this implement/fix? Explain your changes
---------------------------------------------------
This patch create errors.go file and group some thing errors on it 

Where has this been tested?
---------------------------

**Operating System/Platform:** Ubuntu 18.04.4

**Go Version:** 1.13.6

### CODE REVIEW

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- pkg/thing/delivery/http/thing_proxy.go  1
- pkg/thing/handler/amqp/msg_handler.go  2
- pkg/thing/interactors/unregister_thing_test.go  3
- pkg/thing/entities/errors.go  1
- pkg/thing/interactors/unregister_thing.go  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5130340&bid=16859599)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- pkg/thing/entities/errors.go  1
- pkg/thing/interactors/unregister_thing.go  7
- pkg/thing/handler/amqp/msg_handler.go  2
- pkg/thing/delivery/http/thing_proxy.go  1
- pkg/thing/interactors/unregister_thing_test.go  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5130340&bid=16859599)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- pkg/thing/entities/errors.go  1
- pkg/thing/interactors/unregister_thing_test.go  3
- pkg/thing/handler/amqp/msg_handler.go  2
- pkg/thing/interactors/unregister_thing.go  7
- pkg/thing/delivery/http/thing_proxy.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5130340&bid=16859599)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- pkg/thing/handler/amqp/msg_handler.go  2
- pkg/thing/delivery/http/thing_proxy.go  1
- pkg/thing/interactors/unregister_thing_test.go  3
- pkg/thing/interactors/unregister_thing.go  7
- pkg/thing/entities/errors.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5130340&bid=16859599)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- pkg/thing/interactors/unregister_thing.go  7
- pkg/thing/entities/errors.go  1
- pkg/thing/interactors/unregister_thing_test.go  3
- pkg/thing/delivery/http/thing_proxy.go  1
- pkg/thing/handler/amqp/msg_handler.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5130340&bid=16859599)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
           

Complexity increasing per file
==============================
- pkg/thing/interactors/unregister_thing_test.go  3
- pkg/thing/handler/amqp/msg_handler.go  2
- pkg/thing/delivery/http/thing_proxy.go  1
- pkg/thing/entities/errors.go  1
- pkg/thing/interactors/unregister_thing.go  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5130340&bid=16859599)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/delivery/http/thing_proxy.go  1
- pkg/thing/entities/errors.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5130340&bid=16859599)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/delivery/http/thing_proxy.go  1
- pkg/thing/entities/errors.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5130340&bid=16859599)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/entities/errors.go  1
- pkg/thing/delivery/http/thing_proxy.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5130340&bid=16859599)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/entities/errors.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5130340&bid=16859599)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/entities/errors.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5130340&bid=16859599)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/entities/errors.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5130340&bid=16859599)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/entities/errors.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5130340&bid=16859599)

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/register_thing.go  3
- pkg/thing/interactors/request_data.go  2
- pkg/thing/interactors/auth_thing.go  1
         

Clones added
============
- pkg/thing/interactors/unregister_thing_test.go  2
         

Clones removed
==============
+ pkg/thing/interactors/update_schema_test.go  -9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5130340&bid=16859599)

File: pkg/thing/interactors/errors.go [hunk 1]
```diff
@@ -20,4 +22,30 @@ var (
 
 	// ErrSensorsNotProvided error occur when thing's sensors is not provided
 	ErrSensorsNotProvided = errors.New("thing's sensors not provided")
+
+	// ==================== REGISTER THING ERRORS: ====================
```
Comment (lcbm) [inline pkg/thing/interactors/errors.go]: I suggest we remove these comments. All errors in this package are imported with `entities`, the need to comment which interactors use each error (note that some more than one interactor may use a same error) means that we have space for improvement. From my POV, we may either:

1. Further standardize the **error names** and **error comments** (as it is done in the [docker reference file](https://github.com/docker/distribution/blob/master/reference/reference.go#L40)) so that they are generic enough to be reused but specific enough that we know the context which they are used.
>For example, **all** errors which regard to **schema** should be in the format `ErrSchema...`, all **argument** errors should end with `NotProvided` or `Missing`, and so on.

2. Add a subdirectory for each interactor (just like we did we thing/delivery/amqp) and each of those are subpackages with their own `errors.go` file and repeated errors (compiler won't complain since they are unique packages).

Option 1 seems the best for now, as number 2 I'm not sure if its optimal (and it requires changing the project's structure), but from my understanding it seems fine.

**Review by lcbm — COMMENTED** [review]
Awesome :star: 

I Left a _comment_, but it's opinion-based, so let me know your thoughts :smile:

File: pkg/thing/interactors/errors.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+	// ErrNameNotProvided error occur when thing's name is not provided
+	ErrNameNotProvided = errors.New("thing's name not provided")
+
+	// ErrSchemaNotProvided error occur when thing's schema is not provided
+	ErrSchemaNotProvided = errors.New("thing's schema not provided")
+
+	// ErrDataNotProvided error occur when thing's data is not provided
+	ErrDataNotProvided = errors.New("thing's data not provided")
+
+	// ErrSensorsNotProvided error occur when thing's sensors is not provided
```
Comment (netoax) [inline pkg/thing/interactors/errors.go]: "aren't provided"

File: pkg/thing/interactors/errors.go (hunk 1 above)
Comment (netoax) [reply pkg/thing/interactors/errors.go]: Definitely +1 on removing the comments and standardizing the error names and comments. Option 2 is a valid one, of course, but at least for me, it's an over-optimization that leads to an ugly directory tree and the result would be completely the opposite of improving our code.

**Review by netoax — CHANGES_REQUESTED** [review]
_5b2808c79eb140299be75018e854b79ae3f92b10_
This patch group all validating test errors **among** interactors in
errors.go file.

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/request_data.go  2
- pkg/thing/interactors/register_thing.go  3
- pkg/thing/interactors/auth_thing.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5130340&bid=16859599)

File: pkg/thing/interactors/errors.go (hunk 1 above)
Comment (esdrasjnr) [reply pkg/thing/interactors/errors.go]: Done! Agree with @joaoaneto and I adopted the first standard you mentioned @lcbm

File: pkg/thing/interactors/errors.go (hunk 2 above)
Comment (esdrasjnr) [reply pkg/thing/interactors/errors.go]: @joaoaneto Done!

File: pkg/thing/delivery/http/thing_proxy.go [hunk 3]
```diff
@@ -179,7 +179,7 @@ func (p proxy) Get(authorization, ID string) (*entities.Thing, error) {
 		}
 	}
 
-	return nil, &entities.ErrThingNotFound{ID: ID}
+	return nil, entities.ErrThingNotFound
```
Comment (netoax) [inline pkg/thing/delivery/http/thing_proxy.go:182]: There is one thing we need to be aware of before changing the error pattern, some of them were carrying contextual information about the error like this one (which was informing that a specific device couldn't be found). How can we add this information? It was because of these situations that you suggested helper functions to create them, something link `NewErrThingNotFound(id)`?

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/register_thing.go  3
- pkg/thing/interactors/request_data.go  2
- pkg/thing/interactors/auth_thing.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5130340&bid=16859599)

**Review by netoax — CHANGES_REQUESTED** [review]

File: pkg/thing/interactors/errors.go [hunk 4]
```diff
@@ -20,4 +20,22 @@ var (
 
 	// ErrSensorsNotProvided is returned when thing's sensors are not provided
 	ErrSensorsNotProvided = errors.New("thing's sensors not provided")
+
+	// ErrIDLength is returned when the thing's id have more than 16 ascii characters
+	ErrIDLength = errors.New("id length exceeds 16 characters")
+
+	// ErrIDNotInHex is returned when the thing's id is not formatted in hexadecimal base
+	ErrIDNotInHex = errors.New("id is not in hexadecimal format")
```
Comment (netoax) [inline pkg/thing/interactors/errors.go]: `ErrIDNotHex` sounds better to me. What do you think, @lcbm and @esdrasjnr?

File: pkg/thing/interactors/auth_thing.go [hunk 5]
```diff
@@ -11,15 +13,16 @@ func (i *ThingInteractor) Auth(authorization, id string) error {
...
 	_, err := i.thingProxy.Get(authorization, id)
 	if err != nil {
-		i.logger.Error(err)
-		err = i.clientPublisher.SendAuthStatus(id, err)
-		return err
+		sendErr := i.clientPublisher.SendAuthStatus(id, err)
+		if sendErr != nil {
+			return fmt.Errorf("error getting thing metadata: %v: %w", sendErr, err)
+		}
+		return fmt.Errorf("error getting thing metadata: %w", err)
```
Comment (netoax) [inline pkg/thing/interactors/auth_thing.go:20]: Is this change related to grouping business logic errors?

File: pkg/thing/interactors/auth_thing.go [hunk 6]
```diff
@@ -11,15 +13,16 @@ func (i *ThingInteractor) Auth(authorization, id string) error {
...
 	}
 
 	err = i.clientPublisher.SendAuthStatus(id, nil)
 	if err != nil {
-		i.logger.Error(err)
-		return err
+		return fmt.Errorf("error sending response to client: %w", err)
```
Comment (netoax) [inline pkg/thing/interactors/auth_thing.go:25]: Same here.

**Review by netoax — CHANGES_REQUESTED** [review]

File: pkg/thing/interactors/errors.go (hunk 4 above)
Comment (lcbm) [reply pkg/thing/interactors/errors.go]: Agreed @joaoaneto! I don't think we need to use prepositions when declaring identifiers :smile:

File: pkg/thing/delivery/http/thing_proxy.go (hunk 3 above)
Comment (esdrasjnr) [reply pkg/thing/delivery/http/thing_proxy.go:182]: @joaoaneto Since the ID is in the calling context, I think there is no problem to ommit the ID in the returned error. Who gets the error will know what ID is relationed to error.

I suggested helper funcions to help adding contextual info to errors, for example where we are using `fmt.Errorf` to wrap two errors, we could create a function that nest this wrap for us.

File: pkg/thing/interactors/errors.go (hunk 4 above)
Comment (esdrasjnr) [reply pkg/thing/interactors/errors.go]: Done!

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/request_data.go  2
- pkg/thing/interactors/auth_thing.go  1
- pkg/thing/interactors/register_thing.go  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5130340&bid=16859599)

File: pkg/thing/interactors/auth_thing.go (hunk 5 above)
Comment (esdrasjnr) [reply pkg/thing/interactors/auth_thing.go:20]: You're right! I've split it in another commit

File: pkg/thing/interactors/auth_thing.go (hunk 6 above)
Comment (esdrasjnr) [reply pkg/thing/interactors/auth_thing.go:25]: Done!

**Comment by joaoaneto** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/register_thing.go  3
- pkg/thing/interactors/auth_thing.go  1
- pkg/thing/interactors/request_data.go  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5130340&bid=16859599)

File: pkg/thing/interactors/list_things.go [hunk 7]
```diff
@@ -1,13 +1,9 @@
...
 
-import (
-	"errors"
-)
-
 // List fetchs the registered things and return them as an array
 func (i *ThingInteractor) List(authorization string) error {
 	if authorization == "" {
-		return errors.New("authorization key not provided")
+		return ErrAuthNotProvided
```
Comment (netoax) [inline pkg/thing/interactors/list_things.go:8]: Here you're just moving the error creation to another file and not necessarily adding missing argument errors, that's why I suggest you rename your commit title to a more generic one. Maybe it's a good sign to break this commit into two: add missing argument errors and moving argument errors to a unique file.

File: pkg/thing/interactors/list_things_test.go [hunk 8]
```diff
@@ -62,10 +62,10 @@ var ltCases = []listThingsTestCase{
 	{
 		"failed to publish list things response",
 		"authorization key",
-		errors.New("Message queue unavailable"),
+		errors.New("message queue unavailable"),
 		[]*entities.Thing{},
 		nil,
-		errors.New("Message queue unavailable"),
```
Comment (netoax) [inline pkg/thing/interactors/list_things_test.go:68]: Add in the commit message that you're fixing this too.

**Review by netoax — CHANGES_REQUESTED** [review]
_6752e704fcabbe0c693e55f90614a23f9e21bf66_
"In order to standardize the errors in things folder, this patch adds a
file named error.go to define the **common** errors used inside the folder."

File: pkg/thing/interactors/list_things.go (hunk 7 above)
Comment (esdrasjnr) [reply pkg/thing/interactors/list_things.go:8]: You're right, I've split in two commits. Done!

File: pkg/thing/interactors/list_things_test.go (hunk 8 above)
Comment (esdrasjnr) [reply pkg/thing/interactors/list_things_test.go:68]: I putted a message body to describe this :smile:

**Comment by netoax** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pkg/thing/interactors/register_thing.go  3
- pkg/thing/interactors/request_data.go  2
- pkg/thing/interactors/auth_thing.go  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/joaoaneto/knot-babeltower/pullRequest?prid=5130340&bid=16859599)

**Review by netoax — APPROVED** [review]
LGTM!

**Review by lcbm — APPROVED** [review]
LGTM! :confetti_ball:
